edge-modules/ModuleRestarter/src/Settings.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace ModuleRestarter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Configuration;
+    using Newtonsoft.Json;
+    using Newtonsoft.Json.Serialization;
+
+    [JsonObject(NamingStrategyType = typeof(CamelCaseNamingStrategy))]
+    public class Settings
+    {
+        static readonly Lazy<Settings> DefaultSettings = new Lazy<Settings>(
+            () =>
+            {
+                IConfiguration configuration = new ConfigurationBuilder()
+                    .SetBasePath(Directory.GetCurrentDirectory())
+                    .AddJsonFile("config/settings.json", optional: true)
+                    .AddEnvironmentVariables()
+                    .Build();
+
+                return new Settings(
+                    configuration.GetValue<string>("ServiceClientConnectionString"),
+                    configuration.GetValue<string>("IOTEDGE_DEVICEID"),
+                    configuration.GetValue<string>("DesiredModulesToRestartCSV"),
+                    configuration.GetValue<int>("RestartIntervalInMins", 10));
+            });
+
+        Settings(
+            string serviceClientConnectionString,
+            string deviceId,
+            string desiredModulesToRestartCSV,
+            int restartIntervalInMins)
+        {
+            this.ServiceClientConnectionString = Preconditions.CheckNonWhiteSpace(serviceClientConnectionString, "ServiceClientConnectionString");
+            this.DeviceId = Preconditions.CheckNonWhiteSpace(deviceId, "DeviceId");
+            this.RestartIntervalInMins = Preconditions.CheckRange(restartIntervalInMins, 0);
+
+            // mitigate unintended repeated commas
+            this.DesiredModulesToRestart = new List<string>();
+            foreach (string name in desiredModulesToRestartCSV.Split(","))
+            {
+                if (!string.IsNullOrWhiteSpace(name))
+                {
+                    this.DesiredModulesToRestart.Add(name);
+                }
+            }
+        }
+
+        public static Settings Current => DefaultSettings.Value;
+
+        public string ServiceClientConnectionString { get; }
+
+        public string DeviceId { get; }
+
+        public List<string> DesiredModulesToRestart { get; }
+
+        public int RestartIntervalInMins { get; }
+
+
+        public override string ToString()
+        {
+            return JsonConvert.SerializeObject(this, Formatting.Indented);
```

Comment:
why not just construct all properties delimited by new line?

Reply:
I thought this was cleaner than attempting manually. If a new field is added, you do not have to edit the ToString() method. Also, loadgen currently uses the same JsonConvert approach.

Reply:
ServiceClientConnectionString/any secrets should not be printed out without mask.  It is convenient but it may not be exactly your intent (when someone added a new field later).  Also, I prefer simple concatenation as it should be faster than object serialization.

Reply:
I liked the output format with output serialization, as well as the maintainability. Since we can add the tag [JsonIgnore], I prefer the serialization solution.